### PR TITLE
Constant propagation of CProverString.setLength() and CProverString.setCharAt()

### PR DESCRIPTION
**Pull request by danpoe**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: jbmc/regression/jbmc-strings/ConstantEvaluationCProverStringSetCharAt/Test.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import org.cprover.CProverString;
+
+public class Test
+{
+  public void testSuccess1()
```
Comment (smowton) [inline jbmc/regression/jbmc-strings/ConstantEvaluationCProverStringSetCharAt/Test.java:5]: No negative tests? Setting out of range for example?

File: jbmc/regression/jbmc-strings/ConstantEvaluationCProverStringSetLength/Test.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import org.cprover.CProverString;
+
+public class Test
+{
+  public void testSuccess1()
+  {
+    StringBuffer sb = new StringBuffer("abc");
+    CProverString.setLength(sb, 0);
```
Comment (smowton) [inline jbmc/regression/jbmc-strings/ConstantEvaluationCProverStringSetLength/Test.java:8]: Test setting to same length (i.e. 3 in this case)

File: jbmc/regression/jbmc-strings/ConstantEvaluationCProverStringSetLength/Test.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  public void testSuccess2()
+  {
+    StringBuffer sb = new StringBuffer("abc");
+    CProverString.setLength(sb, 2);
+    assert sb.toString().equals("ab");
+  }
+
+  public void testSuccess3()
+  {
+    StringBuffer sb = new StringBuffer("abc");
```
Comment (smowton) [inline jbmc/regression/jbmc-strings/ConstantEvaluationCProverStringSetLength/Test.java:21]: Test extension by more than one char

**Review by smowton — APPROVED** [review]
Looks good, few missing test cases

File: jbmc/regression/jbmc-strings/ConstantEvaluationCProverStringSetCharAt/Test.java (hunk 1 above)
Comment (danpoe) [reply jbmc/regression/jbmc-strings/ConstantEvaluationCProverStringSetCharAt/Test.java:5]: I've not included test cases for the error cases here as those will be handled by the models in the future. I'll add tests for those cases once we have models that use the CProverString primitives.

File: jbmc/regression/jbmc-strings/ConstantEvaluationCProverStringSetLength/Test.java (hunk 2 above)
Comment (danpoe) [reply jbmc/regression/jbmc-strings/ConstantEvaluationCProverStringSetLength/Test.java:8]: Done

File: jbmc/regression/jbmc-strings/ConstantEvaluationCProverStringSetLength/Test.java (hunk 3 above)
Comment (danpoe) [reply jbmc/regression/jbmc-strings/ConstantEvaluationCProverStringSetLength/Test.java:21]: Done

File: jbmc/regression/jbmc-strings/ConstantEvaluationCProverStringSetCharAt/Test.java (hunk 1 above)
Comment (smowton) [reply jbmc/regression/jbmc-strings/ConstantEvaluationCProverStringSetCharAt/Test.java:5]: Sure but we should check that they aren't accidentally const propagated?

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a966852f7392440f251f33dbdaaad1bbca29cd1f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/129478144

File: jbmc/regression/jbmc-strings/ConstantEvaluationCProverStringSetCharAt/Test.java (hunk 1 above)
Comment (danpoe) [reply jbmc/regression/jbmc-strings/ConstantEvaluationCProverStringSetCharAt/Test.java:5]: Fair enough, added more tests.

**Pull request merged**